File: 900_RLE_iterator.java
class RLEIterator {
    int[] A;
    int amountIdx, accumulated;

    public RLEIterator(int[] A) {
        this.A = A;
        amountIdx = accumulated = 0;
    }

    public int next(int n) {
        while (amountIdx < A.length) {
            if (accumulated + n > A[amountIdx]) {
                // because accumulated + n > A[amountIdx]
                // such that A[amountIdx] - accumulated < n
                // we first reduce `accumulated` from A[amountIdx]
                // then subtract n with remaining amount
                n -= (A[amountIdx] - accumulated);

                // with previous subtraction, there is no accumulated elements
                // set to 0
                accumulated = 0;

                // then move amountIdx forward to next amount
                amountIdx += 2;
            } else {
                accumulated += n;
                return A[amountIdx + 1];
            }
        }

        return -1;
    }
}

/**
 * Your RLEIterator object will be instantiated and called as such:
 * RLEIterator obj = new RLEIterator(A);
 * int param_1 = obj.next(n);
 */

File: python/900_RLE_iterator.py
class RLEIterator:

    def __init__(self, A: List[int]):
        self.A = A
        self.idx = 0

    def next(self, n: int) -> int:
        while self.idx < len(self.A) and n > self.A[self.idx]:
            n -= self.A[self.idx]
            self.idx += 2

        if self.idx >= len(self.A):
            return -1

        self.A[self.idx] -= n
        return self.A[self.idx + 1]


# Your RLEIterator object will be instantiated and called as such:
# obj = RLEIterator(A)
# param_1 = obj.next(n)
